### PR TITLE
Do not propagate bad assertions from unreachable blocks

### PR DESCRIPTION
**Pull request by CarolEidt**
and validate that only valid assumptions are utilized.

### CODE REVIEW

**Comment by mikedn** [comment]
Hmm, `GC\Scenarios\DoublinkList\dlstack\dlstack.cmd` failed in one of my PRs as well. It seems unlikely that both my changes and your changes impact that test so there's probably an issue with the test itself.

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib @papaslavik - Please take a look.

File: src/jit/assertionprop.cpp [hunk 1]
```diff
@@ -754,12 +754,13 @@ Compiler::AssertionDsc *  Compiler::optGetAssertion(AssertionIndex assertIndex)
...
+    optDebugCheckAssertion(assertion);
+#endif
 
-    if (assertIndex > optMaxAssertionCount)
-    {
-        return nullptr;
-    }
-    return &optAssertionTabPrivate[assertIndex - 1];
+    return assertion;
 }
```
Comment (CarolEidt) [inline src/jit/assertionprop.cpp]: Note that this previously returned null if the index is outside the bounds of the array, but 1) none of the callers protected against a null result, and 2) this in no way means that it is a valid assertion. Only assertion indices less than or equal to optAssertionCount are valid.

**Comment by pgavlin** [comment]
cc @JosephTremoulet @erozenfeld @AndyAyersMS

**Comment by erozenfeld** [comment]
LGTM

**Comment by CarolEidt** [comment]
`apFull` is used in morph, so that would require additional changes. Perhaps a candidate for future cleanup.

File: src/jit/assertionprop.cpp [hunk 2]
```diff
@@ -1488,6 +1489,63 @@ Compiler::AssertionIndex Compiler::optAddAssertion(AssertionDsc* newAssertion)
...
 
 #ifdef DEBUG
+void Compiler::optDebugCheckAssertion(AssertionDsc* assertion)
+{
+    assert(assertion->assertionKind < OAK_COUNT);
+    assert(assertion->op1.kind < O1K_COUNT);
+    assert(assertion->op2.kind < O2K_COUNT);
+    // It would be good to check that op1.vn and op2.vn are valid value numbers.
+
```
Comment (briansull) [inline src/jit/assertionprop.cpp]: We do have a method that validates a value number:

ValueNumStore::VNIsValid(ValueNum vn)

**Comment by briansull** [comment]
LGTM

**Comment by CarolEidt** [comment]
@briansull @AndyAyersMS - I will submit a new issue for further cleanup (removal of apFull and validation of value numbers in assertions)

**Comment by CarolEidt** [comment]
Add Issue #6492

**Pull request merged**